WIP: Angularify lobby

hnrkndrssn opened the pull request:
This is under :construction: 

Sending PR to start the conversation on this adventure.

This is a :baby_symbol: step towards the :moneybag: 
- [x] Our implementation of internationalization of text doesn't run (as it's no longer a Razor file)
- [x] Join a room
- [x] Search / Filter
- [x] Load more rooms
- [ ] Updating when changes occur
- [ ] Split app/controllers/directives etc into their own files
- [x] Make a directive for the lobby room lists, passing in the room list to use
- [x] Make the room count a function on the scope (remove duplicate HTML in view)
- [x] Rooms list could use an angular filter to split private/public rooms...
- [ ] :question: 

It's a start, let's discuss it and make sure we make it a success :sparkling_heart:

hnrkndrssn committed:
Initial work for angularjs, lobby

Internationalization DOES NOT work at the moment

PureKrome commented:
Can't have an epic JabbR PR/Issue without a meme!

![Nice](https://31.media.tumblr.com/888729e804700b0cd7eb21d1894025a8/tumblr_inline_n4fhthyiuv1solpjm.gif)

I'll do my honorary part.

davidfowl commented on JabbR/Scripts/app/areas/rooms/lobby.html:
```diff
@@ -0,0 +1,47 @@
+                    <span class="topic" data-ng-bind-html="privateRoom.Topic"></span>
+                </div>
+                <div class="span2">
+                    <span class="count" data-ng-show="{{privateRoom.Count == 0}}">@LanguageResources.Client_OccupantsZero</span>
```
This must be a better way of doing this

hnrkndrssn committed:
Make a directive for the room lists in the Lobby

hnrkndrssn committed:
Make the room user count text a function on the view model

cmaish commented on JabbR/Scripts/app/app.js:
```diff
@@ -0,0 +1,71 @@
+    return {
+        restrict: 'A',
+        templateUrl: 'Scripts/app/areas/rooms/lobby-rooms.html',
+        scope: {
```
Doesn't building this as a directive and not a child controller make it harder to move stuff in/down?  Stuff like limiting the room count to 100 until you click "Load more rooms" seems like it'd be pretty awkward when you wrap it into a directive

hnrkndrssn replied:
Refactored the directive to make use of a child controller instead of an isolate scope

cmaish commented on JabbR/Scripts/app/areas/rooms/lobby-rooms.html:
```diff
@@ -0,0 +1,21 @@
+﻿<h2 class="userlist-header nav-header">{{title}}</h2>
+<ul data-empty-message="@LanguageResources.Client_NoMatchingRooms">
+    <li data-ng-repeat="room in rooms" class="room" data-ng-class="{room.Private: locked, room.Closed: closed}" data-room="{{room.Name}}" data-name="{{room.Name}}" data-count="{{room.Count}}">
```
I'd vote for dropping the data prefix - it introduces so much noise for the sake of validation, which doesn't really apply because the attribute values (eg {{room.Name}}) aren't real - they need to be processed to be turned into the desired HTML

cmaish commented on JabbR/Scripts/app/areas/rooms/lobby.html:
```diff
@@ -0,0 +1,5 @@
+﻿<div id="lobby-private" data-ng-show="privateRooms.length > 0" data-jabbr-lobby-rooms data-title="'Your Private Rooms'" data-rooms="privateRooms"></div>
+<div id="lobby-other" data-ng-show="publicRooms.length > 0" data-jabbr-lobby-rooms data-title="'Other Rooms'" data-rooms="publicRooms">
```
I'm thinking we'll need our own ng-bind-esque directive which does localisation and is standalone.  Eg ng-bind-lang

hnrkndrssn replied:
I just added these localisable properties to the scope and use the utilility.getLanguageResource function

hnrkndrssn committed:
Inject our localisation utility to our ng-app

Create child controllers used by the lobby rooms directive to make
extensibility easier

hnrkndrssn committed:
Joining a room

hnrkndrssn committed:
Search and filter room lists

hnrkndrssn committed:
Load more rooms

hnrkndrssn committed:
Use filter to filter private/public rooms

daanl commented on JabbR/Scripts/app/app.js:
```diff
@@ -0,0 +1,100 @@
+    });
+})
+.controller('LobbyController', ['$scope', '$sanitize', '$window', function ($scope, $sanitize, $window) {
+    var connection = $window.jQuery.connection;
```
We should make the hub connection into a dependency

daanl commented on JabbR/Scripts/app/app.js:
```diff
@@ -0,0 +1,100 @@
+})
+.controller('LobbyController', ['$scope', '$sanitize', '$window', function ($scope, $sanitize, $window) {
+    var connection = $window.jQuery.connection;
+    var chat = connection.chat;
```
We should make the chat into a dependency

daanl commented on JabbR/Scripts/app/app.js:
```diff
@@ -0,0 +1,100 @@
+    var chat = connection.chat;
+    var ui = $window.chat.ui;
+    var $ui = $(ui);
+
```
We should make the chat into a dependency, using jquery in a angularjs controller is usually bad practice.
Because we are migrating from plain jquery we maybe need this class for a while until everything is angularjs.

daanl commented on JabbR/Scripts/app/app.js:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    $scope.joinRoom = function (event, room) {
+        console.log('Joining room: ' + room.Name);
```
We should make a logger or use https://docs.angularjs.org/api/ng/service/$log

daanl commented on JabbR/Scripts/app/app.js:
```diff
@@ -0,0 +1,100 @@
+                        console.log(value);
+                        value.getUserCount = function () {
+                            if (this.Count === 0)
+                                return $window.chat.utility.getLanguageResource('Client_OccupantsZero');
```
This is viewlogic shouldn't be in a controller

daanl commented on JabbR/Scripts/app/app.js:
```diff
@@ -0,0 +1,100 @@
+}])
+.controller('LobbyPublicRoomsController', ['$scope', '$window', '$filter', function($scope, $window, $filter) {
+    $scope.isPrivate = false;
+    $scope.title = $window.chat.utility.getLanguageResource('Client_OtherRooms');
```
This is all viewlogic we should rip this out

daanl commented on JabbR/Scripts/app/app.js:
```diff
@@ -0,0 +1,100 @@
+                    console.log(rooms.length);
+                    angular.forEach(rooms, function (value, key) {
+                        console.log(value);
+                        value.getUserCount = function () {
```
getUserCount should be an function on the scope (controller/directive)

hnrkndrssn committed:
Remove view logic from controllers

Created on the scope of the directive instead as per PR feedback

hnrkndrssn committed:
Use $log instead of console.log

hnrkndrssn committed:
Make our JabbR objects dependencies

hnrkndrssn committed:
JabbR Dependencies, take 2

hnrkndrssn commented:
Closing this for now, will revisit at a later date.
